encoding/x/protobuf/inbound.go

```diff
@@ -128,25 +125,83 @@ func newOnewayHandler(
 }
 
 func (o *onewayHandler) HandleOneway(ctx context.Context, transportRequest *transport.Request) error {
-	if err := encoding.Expect(transportRequest, Encoding); err != nil {
+	ctx, _, request, err := getProtoRequest(ctx, transportRequest, o.newRequest)
+	if err != nil {
 		return err
 	}
+	return o.handleOneway(ctx, request)
+}
+
+func getProtoRequest(ctx context.Context, transportRequest *transport.Request, newRequest func() proto.Message) (context.Context, *apiencoding.InboundCall, proto.Message, error) {
+	if err := encoding.Expect(transportRequest, Encoding, JSONEncoding); err != nil {
+		return nil, nil, nil, err
+	}
 	ctx, call := apiencoding.NewInboundCall(ctx)
 	if err := call.ReadFromRequest(transportRequest); err != nil {
-		return err
+		return nil, nil, nil, err
 	}
+	request := newRequest()
+	if err := unmarshal(transportRequest.Encoding, transportRequest.Body, request); err != nil {
+		return nil, nil, nil, encoding.RequestBodyDecodeError(transportRequest, err)
+	}
+	return ctx, call, request, nil
+}
+
+func unmarshal(encoding transport.Encoding, reader io.Reader, message proto.Message) error {
 	buf := buffer.Get()
 	defer buffer.Put(buf)
-	if _, err := buf.ReadFrom(transportRequest.Body); err != nil {
+	if _, err := buf.ReadFrom(reader); err != nil {
```

Comment:
we only need all this buffer decoding for the proto use case.  Since in the JSON case we're converting this byte array back into a reader.  We should move all this buffer/reader interaction into the `unmarshalProto` function and have the `unmarshallJSON` function pass the reader through.